packages/portfolio/src/balance-manager.ts

```diff
@@ -0,0 +1,331 @@
+import {App, Portfolio} from '@yoroi/types'
+import {
+  hasEntryValue,
+  hasValue,
+  observerMaker,
+  queueTaskMaker,
+} from '@yoroi/common'
+import {freeze} from 'immer'
+import {filter} from 'rxjs'
+
+import {sortTokenBalances} from './helpers/sorting'
+import {TokenInfoAndDiscovery} from './types'
+import {isEventTokenManagerSync} from './validators/token-manager-event-sync'
+import {isFt} from './helpers/is-ft'
+import {isNft} from './helpers/is-nft'
+
+export const portfolioBalanceManagerMaker = (
+  {
+    tokenManager,
+    primaryToken,
+    storage,
+    sourceId,
+  }: {
+    tokenManager: Portfolio.Manager.Token
+    storage: Portfolio.Storage.Balance
+    primaryToken: TokenInfoAndDiscovery
+  } & Portfolio.Event.SourceId,
+  {
+    observer = observerMaker<Portfolio.Event.BalanceManager>(),
+    queue = queueTaskMaker(),
+  }: {
+    observer?: App.ObserverManager<Portfolio.Event.BalanceManager>
+    queue?: App.QueueTaskManager
+  } = {},
+): Portfolio.Manager.Balance => {
+  let isHydrated = false
+  let secondaries: Readonly<Map<Portfolio.Token.Id, Portfolio.Token.Balance>> =
+    freeze(new Map())
+  let sortedBalances: Readonly<{
+    all: ReadonlyArray<Portfolio.Token.Balance>
+    nfts: ReadonlyArray<Portfolio.Token.Balance>
+    fts: ReadonlyArray<Portfolio.Token.Balance>
+  }> = freeze({
+    all: [],
+    nfts: [],
+    fts: [],
+  })
+  let primaryBreakdown: Readonly<Portfolio.BalancePrimaryBreakdown> = freeze(
+    {
+      info: primaryToken.info,
+      balance: 0n,
+      lockedInBuiltTxs: 0n,
+      minRequiredByTokens: 0n,
+      records: [],
+    },
+    true,
+  )
+
+  const sort = sortTokenBalancesWrapper({
+    primaryTokenInfo: primaryToken.info,
+  })
+  const updatePrimary = primaryTokenUpdaterWrapper({
+    primaryTokenInfo: primaryToken.info,
+    storagePrimaryBalanceBreakdown: storage.primaryBalanceBreakdown,
+  })
+
+  const subscription = tokenManager.observable
+    .pipe(
+      filter(() => isHydrated),
+      filter((dtoEvent) => isNotTriggeredBySelf(sourceId)(dtoEvent)),
+      filter(
+        (dtoEvent) =>
+          isEventTokenManagerSync(dtoEvent) &&
+          hasStaleTokenInfo(secondaries)(dtoEvent),
```

Comment:
it should also check the hash, yet, since it doesn't load 2 wallets at same time is safe for now.